### PR TITLE
Host targeted refresh deleting and re-creating storage, switch, and lan records

### PR DESCRIPTION
**Pull request by agrare**
https://bugzilla.redhat.com/show_bug.cgi?id=1669176

### CODE REVIEW

File: app/models/manageiq/providers/redhat/inventory/collector/target_collection.rb [hunk 1]
```diff
@@ -213,7 +213,7 @@ def infer_related_host_ems_refs_db!
     changed_hosts.each do |host|
       add_simple_target!(:ems_clusters, uuid_from_target(host.ems_cluster))
       host.storages.each do |storage|
-        add_simple_target!(:storagedomains, uuid_from_target(storage))
+        add_simple_target!(:storagedomains, storage.ems_ref)
```
Comment (agrare) [inline app/models/manageiq/providers/redhat/inventory/collector/target_collection.rb]: This was comparing the wrong property:

```
references(:storagedomains)
["cd4cf9b7-e1d4-4ef6-a3ed-06adf72c8031"]
Storage.pluck(:ems_ref)
["/api/storagedomains/072fbaa1-08f3-4a40-9f34-a5ca22dd1d74", "/api/storagedomains/cd4cf9b7-e1d4-4ef6-a3ed-06adf72c8031"]
```

**Comment by agrare** [comment]
@borod108 the switches reference is empty in the specs, it looks like we get the switch targets and the switches in the parser two different ways.

File: spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_graph_target_host_spec.rb [hunk 2]
```diff
@@ -1,4 +1,8 @@
+require_relative 'ovirt_refresher_spec_common'
+
 describe ManageIQ::Providers::Redhat::InfraManager::Refresh::Refresher do
+  include OvirtRefresherSpecCommon
+
   describe ManageIQ::Providers::Redhat::InfraManager::Refresh::Refresher do
```
Comment (agrare) [inline spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_graph_target_host_spec.rb]: @borod108 why are these describe blocks duplicated?

**Comment by jerryk55** [comment]
@agrare @borod108 @roliveri Just as background, the issue described in https://bugzilla.redhat.com/show_bug.cgi?id=1669176 appears to be fixed by @agrare's change to the storage refresh, and backing out the removal of the network refresh from https://github.com/ManageIQ/manageiq-providers-ovirt/pull/316.  Seeing if the specs pass now...

**Comment by mwperina** [comment]
Moti, could you please review?
@masayag

File: spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_graph_target_host_spec.rb (hunk 2 above)
Comment (borod108) [reply spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_graph_target_host_spec.rb]: For no good reason I can think off.

**Comment by agrare** [comment]
@borod108 yeah I fixed the storages but the networks are still an issue.  I tried checking what `host.network_attachments` returns [here](https://github.com/ManageIQ/manageiq-providers-ovirt/blob/c395e4110ce82c19c8837333ccb79c1fabeebf23/app/models/manageiq/providers/redhat/inventory/collector/target_collection.rb#L224) but in the specs it didn't return anything.

**Comment by jerryk55** [comment]
@agrare @borod108 @roliveri unfortunately my last commit, while fixing the issue tracked here, results in the same error message that https://bugzilla.redhat.com/show_bug.cgi?id=1658240 fixed.  I'm not sure if that's worse that the issue fixed by this PR or not.

**Comment by mwperina** [comment]
Does anyone have an environment to quickly verify this change so we can merge?

File: spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_graph_target_host_spec.rb [hunk 3]
```diff
@@ -42,29 +44,29 @@ def load_response_mock_for(filename)
...
+      expect(serialize_inventory(models_for_host_target)).to eq(saved_inventory)
+
       EmsRefresh.refresh(host)
       host.reload
+
       expect(host.switches.map { |switch| [switch.uid_ems, switch.name] }).to contain_exactly(
         a_collection_containing_exactly("00000000-0000-0000-0000-000000000009", "ovirtmgmt"),
         a_collection_containing_exactly("5c42817f-03fb-460e-a3ab-a8770553aeee", "vlan123t")
       )
-      expect(host.networks.count).to eq(2)
```
Comment (borod108) [inline spec/models/manageiq/providers/redhat/infra_manager/refresh/refresher_graph_target_host_spec.rb]: Just wondering, if we removed this maybe we should also remove the second refresh - (line 63-64)

**Comment by masayag** [comment]
> Does anyone have an environment to quickly verify this change so we can merge?

@mwperina, @borod108 uploaded a commit to this branch that incorporated the proposed changes and updated the test as well.

 ATM the spec are failing, but the code fixes the issue reported by @agrare [here](https://github.com/ManageIQ/manageiq-providers-ovirt/pull/336#issuecomment-459731862)

File: app/models/manageiq/providers/redhat/inventory/collector/target_collection.rb [hunk 4]
```diff
@@ -213,16 +213,21 @@ def infer_related_host_ems_refs_db!
...
+      end
+      host.switches.each do |switch|
+        add_simple_target!(:networks, switch.uid_ems)
```
Comment (agrare) [inline app/models/manageiq/providers/redhat/inventory/collector/target_collection.rb]: This is why the specs are currently failing, this should be `add_simple_target!(:switches, switch.uid_ems)`

File: app/models/manageiq/providers/redhat/inventory/collector/target_collection.rb [hunk 5]
```diff
@@ -213,16 +213,21 @@ def infer_related_host_ems_refs_db!
...
   end
 
   def infer_related_host_ems_refs_api!
     hosts.each do |host|
       add_simple_target!(:ems_clusters, host.cluster.id)
-      host.network_attachments.each do |attachement|
-        add_simple_target!(:networks, attachement.network.id)
+      manager.with_provider_connection(VERSION_HASH) do |connection|
+        connection.follow_link(host.network_attachments).each do |attachement|
+          add_simple_target!(:networks, attachement.network.id)
```
Comment (agrare) [inline app/models/manageiq/providers/redhat/inventory/collector/target_collection.rb]: Same here

**Comment by agrare** [comment]
Okay @masayag @borod108 I pushed fixes for the manager_ref as a hash failures, now it is failing on vcr errors and will have to be re-recorded.

**Comment by jerryk55** [comment]
@agrare the last commit you made still causes the V2V Infrastructure Mapping issue for  missing Networks.

**Comment by agrare** [comment]
@jerryk55 one step at a time :D

**Comment by agrare** [comment]
Okay the issue was that we were still referencing networks when building the targeted arel: https://github.com/ManageIQ/manageiq-providers-ovirt/pull/336/files#diff-c9d64b2c3fed9ede6e26aefcbcb8b276R11

**Comment by agrare** [comment]
@masayag @borod108 it looks like this might need a re-record of the VCRs (I'm not sure why since we didn't change the collectors/parsers at all!?)

**Comment by agrare** [comment]
Testing against a live rhev manager seems to indicate that this is fixed the issue, I'm seeing the same switches and lans before and after host targeted refreshes so I think this just needs a vcr re-record for the specs to pass.

**Comment by jerryk55** [comment]
FYI - still seeing the issue.  not sure if the "one step at a time" comment is still in play...

**Comment by agrare** [comment]
Hey @jerryk55 I tried to reproduce what I think the issue is on a live system and didn't see the problem anymore.
The problem being after a targeted refresh of a host duplicate switches, lans, and storages were created and then the originals deleted after another full refresh.

What issue are you still seeing ?

**Comment by jerryk55** [comment]
Hi @agrare.  Creating an Infrastructure Mapping, then running the RHEV host refresh.  The Infrastructure mapping shows the Networks missing.  Datastore and Cluster are ok.  Prior to your change the Datastore was also showing as missing.  We had debugged this a while back and saw that the ids for the networks were changed causing the issue.

**Comment by agrare** [comment]
```
>> EmsRefresh.refresh(ems)
>> switches = Switch.pluck(:id, :name, :uid_ems)
   (0.6ms)  SELECT "switches"."id", "switches"."name", "switches"."uid_ems" FROM "switches"
=> [[1, "NFS", "ed64724b-4ece-488b-ac9e-3bcfabe09323"], [2, "ovirtmgmt", "00000000-0000-0000-0000-000000000009"]]
>> lans = Lan.pluck(:id, :name, :uid_ems)
   (0.6ms)  SELECT "lans"."id", "lans"."name", "lans"."uid_ems" FROM "lans"
=> [[1, "NFS", "ed64724b-4ece-488b-ac9e-3bcfabe09323"], [2, "ovirtmgmt", "00000000-0000-0000-0000-000000000009"]]
>> EmsRefresh.refresh(ems.hosts.first)
>> switches == Switch.pluck(:id, :name, :uid_ems)
=> true
>> lans == Lan.pluck(:id, :name, :uid_ems)
=> true
```

Looks the same to me, @jerryk55 is this on your local env or do you have a link to an appliance where you see this?

**Comment by jerryk55** [comment]
@agrare link sent in gitter.

**Comment by borod108** [comment]
@agrare any chance can you squash the commits and force push?

**Comment by agrare** [comment]
@borod108 okay I squashed the multiple vcr commits, I think the rest are helpful.

**Comment by agrare** [comment]
@borod108 / @mwperina can you take a look?

**Comment by agrare** [comment]
Thanks @masayag !

**Comment by simaishi** [comment]
Hammer backport details:

```
$ git log -1
commit 01630df8241d85cc5030cb8daa958b61cbc8872b
Author: Moti Asayag <masayag@redhat.com>
Date:   Fri Feb 8 20:43:38 2019 +0200

    Merge pull request #336 from agrare/bz_1669176_host_targeted_refresh
    
    Host targeted refresh deleting and re-creating storage, switch, and lan records
    
    (cherry picked from commit 2509eda7fb579da4c9f31653a04a2d2269e20931)
    
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1674673


```

**Pull request merged**